src/compose/ComposeBox.js

```diff
@@ -391,23 +395,24 @@ class ComposeBoxInner extends PureComponent<Props, State> {
       const topic = this.state.topic.trim();
       return topicNarrow(streamName, topic || '(no topic)');
     }
+    invariant(isConversationNarrow(narrow), 'destination narrow must be conversation');
     return narrow;
   };
 
   handleSend = () => {
     const { dispatch } = this.props;
     const { message } = this.state;
-    const narrow = this.getDestinationNarrow();
+    const destinationNarrow = this.getDestinationNarrow();
```

Comment:
The thing about that is that it'd mean we have two different state properties `topic` and `destinationNarrow` that need to stay in sync -- there'd be an invariant that there needs to be a certain relationship between them. So any place where we updated one, we'd need to be sure to appropriately update the other at the same time.

That certainly can be done, but it's something that can easily acquire subtle bugs. So I'd keep the function until we have a good reason to avoid it.

In a future where we determine the destination narrow in some other way, the function also keeps it flexible for us to change that: its callsites just keep calling it. It might in the future refer to some other piece of state (like a current message, with the idea in your main comment above), or to a `state.destinationNarrow`. If we end up doing the latter, it'd be an easy followup to replace the function calls with direct references to the state then.

Reply:
> it'd mean we have two different state properties `topic` and `destinationNarrow` that need to stay in sync

I agree that this would get tricky. But when the user is composing a PM, in some sense it's a bug that `ComposeBox` is maintaining a `topic` state at all. (In the same way that it would be a bug to maintain a PM-recipients state when composing a stream message.) I'd been thinking—see https://github.com/zulip/zulip-mobile/pull/4773#discussion_r645124186 —that `state.topic` could basically be replaced in `ComposeBox` by a `state.destinationNarrow`. If we need lower-level management of the topic input's state (e.g., for validation that topics aren't empty), we could split off a child component for that, that shows the topic input exactly when we're composing a stream message. I guess extracting that new child component would make even more sense if we're planning to have a counterpart for handling a PM-recipients input for composing PMs.

Thinking some more about this, it occurs to me that `this.state.destinationNarrow` vs. `this.getDestinationNarrow()` will basically collapse once this becomes a Hooks component:

```js
// (if these end up lingering in `ComposeBox` and not split off to new child components)
const [ topic, setTopic ] = useState(/* ... */);
const [ pmRecipients, setPmRecipients ] = useState(/* ... */);

const destinationNarrow = /* ... */;

// Then it's very easy to react to changes in `destinationNarrow` in a
// `useEffect`, e.g., to send typing-stop and typing-start notifications as
// the destination narrow changes
```

So I think after all we don't have a strong need to make a `state.destinationNarrow` now.

Reply:
Yeah. Another angle is to say that for any future where we change what's in the state, if that makes something like `state.destinationNarrow` make more sense than `getDestinationNarrow`, then we can make that change then. It won't be any harder of a refactor than it would be now, and that way we'll only do it if we're confident we'll actually have a need for it.